datafusion-federation/src/sql/ast_analyzer/rewrite_multi_part.rs

```diff
@@ -0,0 +1,526 @@
+use std::ops::ControlFlow;
+
+use datafusion::{
+    common::HashMap,
+    sql::{
+        sqlparser::ast::{self, Ident, ObjectName, VisitMut, VisitorMut},
+        TableReference,
+    },
+};
+
+use crate::sql::table_reference::{MultiPartTableReference, MultiTableReference};
+
+/// Rewrites sqlparser AST statements to use the original table name if any are MultiTableReferences (i.e. "a"."b"."c"."d")
+///
+/// Does nothing if there are no MultiTableReferences.
+#[derive(Debug)]
+pub struct RewriteMultiTableReference;
+
+impl RewriteMultiTableReference {
+    pub fn rewrite(
+        statement: &mut ast::Statement,
+        known_rewrites: HashMap<TableReference, MultiPartTableReference>,
+    ) {
+        let mut known_rewrites_multi = HashMap::new();
+
+        // Iterate over the known rewrites and collect only the multi table references that need rewriting.
+        for (table_reference, multi_table_reference) in known_rewrites {
+            if let MultiPartTableReference::Multi(multi_table_ref) = multi_table_reference {
+                known_rewrites_multi.insert(
+                    table_reference_to_object_name(&table_reference),
+                    multi_table_ref,
+                );
+            }
+        }
+
+        if known_rewrites_multi.is_empty() {
+            return;
+        }
+
+        let mut visitor = RewriteMultiTableVisitor {
+            known_rewrites: known_rewrites_multi,
+        };
+
+        VisitMut::visit(statement, &mut visitor);
+    }
+}
+
+struct RewriteMultiTableVisitor {
+    known_rewrites: HashMap<ObjectName, MultiTableReference>,
+}
+
+impl VisitorMut for RewriteMultiTableVisitor {
+    type Break = ();
+
+    fn pre_visit_relation(&mut self, table_factor: &mut ObjectName) -> ControlFlow<Self::Break> {
```

Comment:
The only two rewrite we need to do are on table_factors and compound expressions

Reply:
I'm very happy with how simple this ended up being compared to the previous implementation